File: plover_vm/kernel_scenario.py
"""Scenario runner for host-side kernel bring-up (S6)."""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

from kern.kernel import Kernel
from kern.vfdd import VfddDriver
from plover_vm.memory.bus import MemoryBus
from plover_vm.memory.mailbox import CMD_READ, MB_BUFFER, MB_CMD, MB_PARAM
from plover_vm.memory.vfdd import VfdConfig, VirtualFdd


@dataclass
class KernelScenarioResult:
    ok: bool
    output: list[str] = field(default_factory=list)
    error: str | None = None


def run_kernel_scenario(doc: dict) -> KernelScenarioResult:
    bus = MemoryBus()
    k = Kernel(bus)
    try:
        for action in doc.get("actions", []):
            typ = action.get("type")
            if typ == "stage0_load_sector0":
                img = Path(action.get("image", "hw/fixtures/vfdd/dos_boot.img"))
                sectors = int(action.get("sectors", 64))
                dev = VirtualFdd(VfdConfig(path=img, sector_count=sectors))
                drv = VfddDriver(dev)
                # Simulated BIOS mailbox read of sector0.
                bus.mailbox.set_sector_stub(drv.read_sector(0))
                bus.write_cpu(MB_PARAM, 0)
                bus.write_cpu(MB_CMD, CMD_READ)
                # Copy mailbox payload to RAM 0x0800.
                for i in range(248):
                    bus.write_cpu(0x0800 + i, bus.read_cpu(MB_BUFFER + i))
                k.kprint("bios_stage0_ok")
            elif typ == "stage1_gpio_smoke":
                # Bare-metal style: switch input controls LED output bit0.
                sw_on = int(action.get("switch", 1)) & 1
                k.gpio.direction = 0x0F  # low nibble output
                # bit5 is switch input
                k.gpio.set_input_bits(mask=(1 << 5), values=(sw_on << 5))
                if k.gpio.get_bit(5):
                    k.gpio.set_bit(0)
                    k.kprint("gpio_smoke_led_on")
                else:
                    k.gpio.clear_bit(0)
                    k.kprint("gpio_smoke_led_off")
            elif typ == "boot":
                k.boot()
            elif typ == "alloc":
                k.kmalloc(int(action.get("bytes", 0)))
            else:
                raise ValueError(f"unknown kernel action: {typ}")
    except Exception as e:  # noqa: BLE001
        return KernelScenarioResult(ok=False, output=list(k.state.output), error=str(e))

    exp = doc.get("expect", {})
    ok = True
    if "output_contains" in exp:
        for s in exp["output_contains"]:
            if not any(s in line for line in k.state.output):
                ok = False
    return KernelScenarioResult(ok=ok, output=list(k.state.output))

File: crates/plover_os/src/kernel_scenario.rs
use crate::kernel::Kernel;
use crate::vfdd::VfddDriver;
use plover_copro::mailbox::{CMD_READ, MB_BUFFER, MB_CMD, MB_PARAM};
use plover_copro::vfdd::{VfdConfig, VirtualFdd};
use plover_mmu::MemoryBus;
use std::path::Path;

#[derive(Debug, Clone)]
pub struct KernelScenarioResult {
    pub ok: bool,
    pub output: Vec<String>,
    pub error: Option<String>,
}

pub fn run_kernel_scenario_yaml(
    actions: &[serde_yaml::Value],
    expect: &serde_yaml::Value,
    root: &Path,
) -> KernelScenarioResult {
    let bus = MemoryBus::default();
    let mut k = Kernel::new(bus);
    for action in actions {
        let typ = action.get("type").and_then(|v| v.as_str()).unwrap_or("");
        if let Err(e) = apply_kernel_action(&mut k, action, typ, root) {
            return KernelScenarioResult {
                ok: false,
                output: k.state.output.clone(),
                error: Some(e),
            };
        }
    }
    let mut ok = true;
    if let serde_yaml::Value::Mapping(map) = expect {
        if let Some(items) = map.get(&serde_yaml::Value::String("output_contains".into())) {
            if let serde_yaml::Value::Sequence(seq) = items {
                for item in seq {
                    if let serde_yaml::Value::String(s) = item {
                        if !k.state.output.iter().any(|line| line.contains(s)) {
                            ok = false;
                        }
                    }
                }
            }
        }
    }
    KernelScenarioResult {
        ok,
        output: k.state.output,
        error: None,
    }
}

fn apply_kernel_action(
    k: &mut Kernel,
    action: &serde_yaml::Value,
    typ: &str,
    root: &Path,
) -> Result<(), String> {
    match typ {
        "stage0_load_sector0" => {
            let img_rel = action
                .get("image")
                .and_then(|v| v.as_str())
                .unwrap_or("hw/fixtures/vfdd/dos_boot.img");
            let sectors = action
                .get("sectors")
                .and_then(|v| v.as_u64())
                .unwrap_or(64) as usize;
            let path = root.join(img_rel.replace('/', std::path::MAIN_SEPARATOR_STR));
            let dev = VirtualFdd::new(VfdConfig {
                path,
                sector_count: sectors,
            })
            .map_err(|e| e.to_string())?;
            let drv = VfddDriver::new(dev);
            let sector = drv.read_sector(0).map_err(|e| format!("{e:?}"))?;
            k.bus.mailbox.set_sector_stub(&sector);
            k.bus.write_cpu(MB_PARAM, 0);
            k.bus.write_cpu(MB_CMD, CMD_READ);
            for i in 0..248 {
                let b = k.bus.read_cpu(MB_BUFFER.wrapping_add(i as u16));
                k.bus.write_cpu(0x0800u16.wrapping_add(i as u16), b);
            }
            k.kprint("bios_stage0_ok");
        }
        "stage1_gpio_smoke" => {
            let sw_on = action
                .get("switch")
                .and_then(|v| v.as_u64())
                .unwrap_or(1) as u8
                & 1;
            k.gpio.direction = 0x0F;
            k.gpio.set_input_bits(1 << 5, if sw_on != 0 { 1 << 5 } else { 0 });
            if k.gpio.get_bit(5) {
                k.gpio.set_bit(0);
                k.kprint("gpio_smoke_led_on");
            } else {
                k.gpio.clear_bit(0);
                k.kprint("gpio_smoke_led_off");
            }
        }
        "set_slots" => {
            if let Some(slots) = action.get("slots").and_then(|v| v.as_mapping()) {
                for (ks, v) in slots {
                    let slot: u8 = match ks {
                        serde_yaml::Value::String(s) => {
                            s.trim_matches('"').parse().unwrap_or(0)
                        }
                        serde_yaml::Value::Number(n) => n.as_u64().unwrap_or(0) as u8,
                        _ => 0,
                    };
                    let sig: u8 = v
                        .as_u64()
                        .or_else(|| v.as_i64().map(|i| i as u64))
                        .unwrap_or(0xFF) as u8;
                    k.state.slot_signatures.insert(slot, sig);
                }
            }
        }
        "boot" => k.boot(),
        "alloc" => {
            let _n = action.get("bytes").and_then(|v| v.as_u64()).unwrap_or(0);
        }
        other => return Err(format!("unknown kernel action: {other}")),
    }
    Ok(())
}
